### PR TITLE
Svetlana/shallow water pk

### PR DESCRIPTION
**Pull request by stokareva**
New solver for shallow water model

### CODE REVIEW

File: src/pks/shallow_water/ShallowWaterBoundaryFunction.cc [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+ */
+
+#include "WhetStoneDefs.hh"
+
+#include "ShallowWaterBoundaryFunction.hh"
+
+namespace Amanzi {
+    namespace ShallowWater {
+        
+        /* ****************************************************************
```
Comment (lipnikov) [inline src/pks/shallow_water/ShallowWaterBoundaryFunction.cc:19]: No additional spacing inside namespaces.

File: src/pks/shallow_water/ShallowWaterBoundaryFunction.hh [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+#include "Mesh.hh"
+#include "WhetStoneDefs.hh"
+
+#include "PK_DomainFunction.hh"
+
+namespace Amanzi {
+    namespace ShallowWater {
+        
+        class ShallowWaterBoundaryFunction : public PK_DomainFunction {
```
Comment (lipnikov) [inline src/pks/shallow_water/ShallowWaterBoundaryFunction.hh:30]: No additional spacing inside namespace.

File: src/pks/shallow_water/ShallowWater_TI.cc [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+#include "UniqueLocalIndex.hh"
+
+namespace Amanzi {
+namespace ShallowWater {
+    
+ void ShallowWater_PK::FunctionalTimeDerivative(double t, const Epetra_Vector& component,
+                                                Epetra_Vector& f_component) {};
+    
+        
+} // ShallowWater
```
Comment (lipnikov) [inline src/pks/shallow_water/ShallowWater_TI.cc:26]: Be consistent across files: "ShallowWater" or "namespace ShalowWater". I use the later format in my code, I think it complies with Amanzi Codying Standards.

File: src/pks/shallow_water/test/Main.cc [hunk 4]
```diff
@@ -0,0 +1,15 @@
+#include <UnitTest++.h>
+
+#include "Teuchos_GlobalMPISession.hpp"
+
+#include "VerboseObject_objs.hh"
+#include "bilinear_form_registration.hh"
```
Comment (lipnikov) [inline src/pks/shallow_water/test/Main.cc:6]: Is it needed? Do you use WhetStone?

File: src/pks/shallow_water/test/shallow_water_1D.cc [hunk 5]
```diff
@@ -0,0 +1,337 @@
...
+
+#include "ShallowWater_PK.hh"
+
+#include "OutputXDMF.hh"
+
+//--------------------------------------------------------------
+// bottom topography
+//--------------------------------------------------------------
+double Bathymetry(double x, double y) {
+	return 0.;
```
Comment (lipnikov) [inline src/pks/shallow_water/test/shallow_water_1D.cc:31]: Amanzi uses 2 spaces for indentation.

File: src/pks/shallow_water/test/shallow_water_1D.cc [hunk 6]
```diff
@@ -0,0 +1,337 @@
...
+}
+
+void dam_break_1D_setIC(Teuchos::RCP<const Amanzi::AmanziMesh::Mesh> mesh_, Teuchos::RCP<Amanzi::State>& S_) {
+
+	int ncells_owned = mesh_->num_entities(Amanzi::AmanziMesh::CELL, Amanzi::AmanziMesh::Parallel_type::OWNED);
```
Comment (lipnikov) [inline src/pks/shallow_water/test/shallow_water_1D.cc:36]: Amanzi uses 2 spaces for indentation

File: src/pks/shallow_water/test/shallow_water_2D_smooth.cc [hunk 7]
```diff
@@ -0,0 +1,352 @@
...
+
+#include "ShallowWater_PK.hh"
+
+#include "OutputXDMF.hh"
+
+//--------------------------------------------------------------
+// bottom topography
+//--------------------------------------------------------------
+double Bathymetry(double x, double y) {
+	return 0.;
```
Comment (lipnikov) [inline src/pks/shallow_water/test/shallow_water_2D_smooth.cc:32]: Amanzi uses 2 spaces for indentation.

File: src/pks/transport/test/transport_2D.cc [hunk 8]
```diff
@@ -92,6 +92,7 @@ std::cout << "Test: Advance on a 2D square mesh: limiter=" << limiter
 
   AmanziGeometry::Point velocity(1.0, 1.0);
   int nfaces_owned = mesh->num_entities(AmanziMesh::FACE, AmanziMesh::Parallel_type::OWNED);
+  std::cout << "nfaces_owned = " << nfaces_owned << std::endl;
```
Comment (lipnikov) [inline src/pks/transport/test/transport_2D.cc:95]: Is it needed here?

File: src/pks/shallow_water/ShallowWater_PK.cc [hunk 9]
```diff
@@ -0,0 +1,1301 @@
...
+
+#include <vector>
+
+// Amanzi::ShallowWater
+#include "ShallowWater_PK.hh"
+
+namespace Amanzi {
+namespace ShallowWater {
+    
+    ShallowWater_PK::ShallowWater_PK(Teuchos::ParameterList& pk_tree,
```
Comment (lipnikov) [inline src/pks/shallow_water/ShallowWater_PK.cc:20]: Amanzi does not use additional spaces inside namespace

File: src/pks/shallow_water/ShallowWater_PK.cc [hunk 10]
```diff
@@ -0,0 +1,1301 @@
...
+
+             FS.resize(3);
+
+             FNum.resize(3);
+
+             for (int i = 0; i < 3; i++) FS[i] = 0.;
+
+             for (int f = 0; f < cfaces.size(); f++) {
+
+				 int orientation;
```
Comment (lipnikov) [inline src/pks/shallow_water/ShallowWater_PK.cc:152]: Amanzi uses 2 spaces for indentation.

File: src/pks/shallow_water/ShallowWater_PK.cc [hunk 11]
```diff
@@ -0,0 +1,1301 @@
...
+
+        apx = std::max(std::max(std::fabs(uL)+std::sqrt(g*hL),std::fabs(uR)+std::sqrt(g*hR)),0.);
+        apy = std::max(std::max(std::fabs(vL)+std::sqrt(g*hL),std::fabs(vR)+std::sqrt(g*hR)),0.);
+        ap  = std::max(apx,apy);
+
+        amx = std::min(std::min(std::fabs(uL)-std::sqrt(g*hL),std::fabs(uR)-std::sqrt(g*hR)),0.);
+        amy = std::min(std::min(std::fabs(vL)-std::sqrt(g*hL),std::fabs(vR)-std::sqrt(g*hR)),0.);
+        am  = std::min(amx,amy);
+
+//        std::cout << "UL: " << UL[0] << " " << UL[1] << " " << UL[2] << std::endl;
```
Comment (lipnikov) [inline src/pks/shallow_water/ShallowWater_PK.cc:1073]: I think this debug code could be removed.

**Comment by jd-moulton** [comment]
I checked in the developers guide, and currently we don' t point out that we want developers to use our version of the Google C++ coding standard.   So no suprise some of these formatting issues fell through.   Please review the standard, we have it in the subdirectory:

[https://github.com/amanzi/amanzi/tree/master/doc/standards/cxx](https://github.com/amanzi/amanzi/tree/master/doc/standards/cxx)

This doesn't render easily from github, but from the repo of the source you can just load the html file into a browser. We've been a little to lenient on standards compliance, particularly for older code that was being merged in.   We should tighten up on standards now though.

**Comment by lipnikov** [comment]
doc/standards/cxx/cppguide.html should revisited.
For instance, we relaxed this requirement:
"Each line of text in your code should be at most 80 characters long."

File: src/pks/flow/test/flow_richards_2D.cc [hunk 12]
```diff
@@ -40,6 +40,7 @@ TEST(FLOW_2D_RICHARDS) {
 
   Comm_ptr_type comm = Amanzi::getDefaultComm();
   int MyPID = comm->MyPID();
+  std::cout << "MyPID = " << MyPID << std::endl;
```
[the remark this answers is not in the document]
Comment (stokareva) [reply src/pks/flow/test/flow_richards_2D.cc]: Removed.

File: src/pks/shallow_water/ShallowWaterBoundaryFunction.cc [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+ 
+ Author: Svetlana Tokareva (tokareva@lanl.gov)
+ */
+
+#include "WhetStoneDefs.hh"
+
+#include "ShallowWaterBoundaryFunction.hh"
+
+namespace Amanzi {
+    namespace ShallowWater {
```
[the remark this answers is not in the document]
Comment (stokareva) [reply src/pks/shallow_water/ShallowWaterBoundaryFunction.cc]: Fixed format

File: src/pks/shallow_water/ShallowWaterBoundaryFunction.cc (hunk 1 above)
Comment (stokareva) [reply src/pks/shallow_water/ShallowWaterBoundaryFunction.cc]: Fixed

File: src/pks/shallow_water/ShallowWaterBoundaryFunction.hh (hunk 2 above)
Comment (stokareva) [reply src/pks/shallow_water/ShallowWaterBoundaryFunction.hh]: Fixed

File: src/pks/shallow_water/test/shallow_water_1D.cc (hunk 6 above)
Comment (stokareva) [reply src/pks/shallow_water/test/shallow_water_1D.cc:36]: Fixed

File: src/pks/shallow_water/test/shallow_water_2D_smooth.cc (hunk 7 above)
Comment (stokareva) [reply src/pks/shallow_water/test/shallow_water_2D_smooth.cc:32]: Fixed

File: src/pks/transport/test/transport_2D.cc (hunk 8 above)
Comment (stokareva) [reply src/pks/transport/test/transport_2D.cc:95]: Removed

File: src/pks/shallow_water/ShallowWater_PK.cc (hunk 10 above)
Comment (stokareva) [reply src/pks/shallow_water/ShallowWater_PK.cc:152]: Fixed

File: src/pks/shallow_water/ShallowWater_PK.cc (hunk 11 above)
Comment (stokareva) [reply src/pks/shallow_water/ShallowWater_PK.cc:1073]: Removed

File: src/pks/shallow_water/ShallowWater_TI.cc (hunk 3 above)
Comment (stokareva) [reply src/pks/shallow_water/ShallowWater_TI.cc:25]: Fixed

File: src/pks/shallow_water/test/Main.cc (hunk 4 above)
Comment (stokareva) [reply src/pks/shallow_water/test/Main.cc:6]: Deleted unnecessary includes

File: src/pks/shallow_water/test/shallow_water_1D.cc (hunk 5 above)
Comment (stokareva) [reply src/pks/shallow_water/test/shallow_water_1D.cc]: Fixed

File: src/pks/shallow_water/ShallowWater_PK.cc (hunk 9 above)
Comment (stokareva) [reply src/pks/shallow_water/ShallowWater_PK.cc]: Fixed

File: src/pks/shallow_water/ShallowWater_PK.cc [hunk 14]
```diff
@@ -0,0 +1,1301 @@
...
+
+
+
+    //--------------------------------------------------------------
+    // calculation of time step from CFL condition
+    //--------------------------------------------------------------
+    double ShallowWater_PK::get_dt() {
+
+    	double h, u, v, g = 9.81;
+		double S, Smax;
```
[the remark this answers is not in the document]
Comment (stokareva) [reply src/pks/shallow_water/ShallowWater_PK.cc]: Fixed

**Comment by stokareva** [comment]
Please review changes

**Comment by lipnikov** [comment]
Please commit file test/shallow_water_1D_convergence.cc

**Comment by stokareva** [comment]
Done. However this test is not used at the moment.

**Comment by lipnikov** [comment]
The tests require long time. Please reduce the runtime to a few seconds. As an example, see timings for transport tests.

**Comment by lipnikov** [comment]
> Done. However this test is not used at the moment.
However, it failed on Travis CI. If it is not ready, remove it from CMakeLists.txt

**Comment by lipnikov** [comment]
1/2 Test #1: shallow_water_1D .................   55.05 sec
2/2 Test #2: shallow_water_2D_smooth ..........   46.62 sec

Looks better, can we run these tests in parallel?

**Comment by lipnikov** [comment]
Just noticed that Visit output is different is serial and parallel.

**Comment by lipnikov** [comment]
The 2D test crashes in parallel or produces incorrect convergence rate.

**Comment by lipnikov** [comment]
Found a bug: cn < 0 in line 570 of ShallowWater_PK.cc

**Pull request merged**